hashintel/hash
# Dynamically generate types for unseen data, and fix GitHub integration

## 🌟 What is the purpose of this PR?

Quite a while ago we added a preliminary "integration" that was able to ingest data from GitHub (#503). This was able to be ran in development environments, but wasn't part of CI or our deployments.

That development work helped support the need for the development of the Block Protocol Type System, and work was paused on it. This PR represents a small return to the work of integrations, and introduces dynamic generation of types (expressed within the Type System) for unseen JSON data. Utilising this work, it "fixes" the GitHub integration, under the state it was originally introduced in; meaning it will still not be available in deployments, and isn't ready for usage outside of development.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203363157432094/1203657039764067/f) _(internal)_
- [Asana task](https://app.asana.com/0/1203363157432094/1203657039764066/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?

Please see commits

## 📜 Does this require a change to the docs?

- As this is reserved for development, it is currently not well captured in docs. This will be fixed in the future as the code becomes a more stable part of the codebase.

## ⚠️ Known issues

- Running multiple ingests in a row for the same data streams will likely result in unwanted behaviour (potentially crashes), we currently do not check for existing data (types or entities) before trying to create them
- Ingest is very slow, this is partially due to the nature of the Airbyte connectors, but also due to inefficient code written here (again, just for development purposes at the moment) and the lack of a bulk/batch insertion endpoint on the Graph (which means we need a roundabout call for every record)

## 🐾 Next steps

- Implement an Asana connector

## 🛡 What tests cover this?

- None besides lints

## ❓ How to test this?

1. Checkout the branch / view the deployment
1. Follow the README to start the app as normal (including `external-services` and `yarn dev`)
1. Run `yarn workspace @hashintel/hash-task-executor start`
1. Open the front-end and login to generate the appropriate cookies
1. Open the GraphQL playground (http://localhost:5001/graphql)
1. Acquire a [personal access token for GitHub](https://github.com/settings/tokens), the scope will depend on the streams you want to ingest, but I have attached my scopes at the bottom
1. Try the following mutation
   ```graphql
   mutation {
     executeGithubSpecTask
   }
   ```
1. Then you can try the following (replace the repositories, access token, and date as needed)
   ```graphql
   mutation {
     executeGithubCheckTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Then you can run the following to download the "catalog" for the connector (a list of available streams)
   ```graphql
   mutation {
     executeGithubDiscoverTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. The previous mutation should have created a file: `packages/hash/task-executor/src/tasks/source-github/secrets/catalog.json`. Format the file using your IDE, and select a few "streams" that you want to download (delete the ones you do not want). The more you pick, the slower it will be to test. Some streams are not filterable by `start_date` so may take a long time even if you pick a recent `start_date`
1. Once you have modified `catalog.json` as you wish, run the following mutation (again updating parameters as you wish) 
   ```graphql
   mutation {
     executeGithubReadTask(config:{
       repository: "blockprotocol/blockprotocol hashintel/hash"
       start_date: "2023-01-01T00:00:00Z"
       credentials: {
         personal_access_token: ${PERSONAL_ACCESS_TOKEN}
       }
     })
   }
   ```
1. Check the app, you should see new entity types, and going to the relevant entity type page should show entities available.

## 📹 Demo
<img width="319" alt="image" src="https://user-images.githubusercontent.com/25749103/210834505-0e4f53ee-6a1c-4f30-8c34-e9e6e590e844.png">
<img width="1399" alt="image" src="https://user-images.githubusercontent.com/25749103/210834569-ecbbf6c1-9066-4eda-bec7-817456cb8650.png">
<img width="496" alt="image" src="https://user-images.githubusercontent.com/25749103/210834680-a0ed6918-908c-47d0-9087-03d64b135071.png">


## Access Token Scopes
<img width="640" alt="image" src="https://user-images.githubusercontent.com/25749103/210820259-ab4246f2-bb21-4407-9d52-f11b72874e91.png">
<img width="598" alt="image" src="https://user-images.githubusercontent.com/25749103/210820327-9e98111f-00d2-48bf-9979-27590d36e856.png">
